src/coreclr/jit/importercalls.cpp

```diff
@@ -5720,25 +5730,68 @@ void Compiler::pickGDV(GenTreeCall*           call,
     // Prefer class guess as it is cheaper
     if (numberOfClasses > 0)
     {
-        unsigned likelihoodThreshold = isInterface ? 25 : 30;
-        if (likelyClasses[0].likelihood >= likelihoodThreshold)
+        const int maxNumberOfGuesses = min(MAX_GDV_TYPE_CHECKS, JitConfig.JitGuardedDevirtualizationMaxTypeChecks());
+        if (maxNumberOfGuesses == 0)
         {
-            *classGuess = (CORINFO_CLASS_HANDLE)likelyClasses[0].handle;
-            *likelihood = likelyClasses[0].likelihood;
+            // DOTNET_JitGuardedDevirtualizationMaxTypeChecks=0 means we don't want to do any guarded devirtualization
+            // Although, we expect users to disable GDV by setting DOTNET_JitEnableGuardedDevirtualization=0
             return;
         }
 
-        JITDUMP("Not guessing for class; likelihood is below %s call threshold %u\n",
-                isInterface ? "interface" : "virtual", likelihoodThreshold);
+        assert((maxNumberOfGuesses > 0) && (maxNumberOfGuesses <= MAX_GDV_TYPE_CHECKS));
+
+        unsigned likelihoodThreshold;
+        if (maxNumberOfGuesses == 1)
+        {
+            // We're allowed to make only a single guess - it means we want to work only with dominating types
+            likelihoodThreshold = isInterface ? 25 : 30;
+        }
+        else if (maxNumberOfGuesses == 2)
+        {
+            // Two guesses - slightly relax the thresholds
+            likelihoodThreshold = isInterface ? 15 : 20;
+        }
+        else
+        {
+            // We're allowed to make more than 2 guesses - pick all types with likelihood >= 10%
```

Comment:
Note with an 8 entry reservoir the smallest value we'll ever see is 12.5%.

Reply:
I assume it's an implementation detail JIT doesn't know, right? Do you want me to change it to 12.5?

Reply:
No, just pointing out that currently you may never see a candidate that will fail this test.

Ultimately, we might want to compute the conditional probability. Say there are two candidates with likelihoods of 90 and 9. If the target is not the first candidate, the conditional probability of it being the second candidate is now 90%. (and if there was a third that was say 0.9, if it was not the first two, it is 90% likely to be that third one).

But to do this, we'd have to have more confidence in those low likelihoods.